File: Archipelago/sayonarawildhearts/options.py
from dataclasses import dataclass

from Options import Choice, OptionGroup, PerGameCommonOptions, Range, Toggle

class RequiredRank(Choice):
    """
    Rank required to clear a level.
    """
    display_name = "Required Rank"
    option_gold = 1
    option_silver = 2
    option_bronze = 3
    default = 1

class EnableDeathLink(Toggle):
    """
    When you die, everyone who enabled death link dies. Of course, the reverse is true too.
    """
    display_name = "Death Link"

class DeathLinkAmnesty(Range):
    """
    How many deaths it takes to send a DeathLink.
    """
    display_name = "Death Link Amnesty"
    range_start = 1
    range_end = 10
    default = 1

@dataclass
class SayonaraWildHeartsOptions(PerGameCommonOptions):
    RequiredRank: RequiredRank
    EnableDeathLink: EnableDeathLink
    DeathLinkAmnesty: DeathLinkAmnesty

sayonara_wild_hearts_option_groups = [
    OptionGroup("Game Options", [
        RequiredRank,
        EnableDeathLink,
        DeathLinkAmnesty
    ])
]

File: Randomizer/Options.cs
using Newtonsoft.Json.Linq;

namespace SayonaraWildHeartsRandomizer;

public class Options
{
    // Rank required to clear a level.
    public enum ERequiredRank
    {
        Gold = 1,
        Silver = 2,
        Bronze = 3,
    }
    public ERequiredRank RequiredRank = ERequiredRank.Gold;

    // When you die, everyone who enabled death link dies. Of course, the reverse is true too.
    public bool EnableDeathLink = false;

    // How many deaths it takes to send a DeathLink.
    public int DeathLinkAmnesty = 1;

    public void Load(MultiWorld multiWorld)
    {
        JObject options;
        try
        {
            options = (JObject)multiWorld.slotData["Options"];
        }
        catch
        {
            return;
        }

        Plugin.Logger.LogInfo(options.ToString());

        if (options["RequiredRank"] != null)
        {
            RequiredRank = (ERequiredRank)int.Parse(options["RequiredRank"].ToString());
        }

        if (options["EnableDeathLink"] != null)
        {
            EnableDeathLink = int.Parse(options["EnableDeathLink"].ToString()) == 1;
        }

        if (options["DeathLinkAmnesty"] != null)
        {
            DeathLinkAmnesty = int.Parse(options["DeathLinkAmnesty"].ToString());
        }
    }
}
